sycl/include/sycl/ext/oneapi/matrix/matrix-jit.hpp

```diff
@@ -530,211 +376,92 @@ class wi_element<uint16_t, NumRows, NumCols, Layout, Group> {
+                                op make_fp32(rhs))),                           \
+        idx);                                                                  \
+    return *this;                                                              \
+  }
+#else // __SYCL_DEVICE_ONLY__
+#define OP(op)                                                                 \
+  wi_element &operator op##=(const uint16_t &rhs) {                            \
+    (void)rhs;                                                                 \
+    throw runtime_error("joint matrix is not supported on host device.",       \
+                        PI_INVALID_DEVICE);                                    \
+  }
+#endif // __SYCL_DEVICE_ONLY__
+  OP(+)
+  OP(-)
+  OP(*)
+  OP(/)
+#undef OP
+
+  template <typename T1, typename T2> struct Converter {
+    static T2 convert(const T1 &from) { return static_cast<T2>(from); }
+  };
 
-  friend bool
-  operator!=(const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,
-             const uint16_t &rhs) {
-#ifdef __SYCL_DEVICE_ONLY__
-    return std::fabs(
-               make_fp32(__spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) -
-               make_fp32(rhs)) >= std::numeric_limits<float>::epsilon();
-#else
-    (void)lhs;
-    (void)rhs;
-    throw runtime_error("joint matrix is not supported on host device.",
-                        PI_INVALID_DEVICE);
+  template <typename T> struct Converter<T, uint16_t> {
+    static uint16_t convert(const T &from) { return make_bf16(from); }
+  };
+#if __SYCL_DEVICE_ONLY__
+#define OP(input_type, type, op)                                               \
+  friend type operator op(                                                     \
+      const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,        \
+      const uint16_t &rhs) {                                                   \
+    return Converter<input_type, type>::convert(make_fp32(                     \
+        __spirv_VectorExtractDynamic(lhs.M.spvm, lhs.idx)) op make_fp32(rhs)); \
+  }                                                                            \
+  friend type operator op(                                                     \
+      const uint16_t &lhs,                                                     \
+      const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &rhs) {      \
+    return Converter<input_type, type>::convert(make_fp32(                     \
+        __spirv_VectorExtractDynamic(rhs.M.spvm, rhs.idx)) op make_fp32(lhs)); \
+  }
+  OP(float, uint16_t, +)
+  OP(float, uint16_t, -)
+  OP(float, uint16_t, *)
+  OP(float, uint16_t, /)
+  OP(bool, bool, ==)
+  OP(bool, bool, !=)
+  OP(bool, bool, <)
+  OP(bool, bool, >)
+  OP(bool, bool, <=)
+  OP(bool, bool, >=)
+#undef OP
+#else // __SYCL_DEVICE_ONLY__
+#define OP(type, op)                                                           \
```

Comment:
If you add the `input_type` parameter here, even though you won't use it, then you can do like you do in the other cases and only have the uses once.

Reply:
Eh, if i understand correctly, you are saying. we do the following thing. then OP(float, uint16_t, +) can be shared by host and device code, we only need to maintain only one copy of "OP(xxx, uint16_t, xx)". Am i right?
```
// code which defines OP for device but we don't undef OP
#define OP(input type, type, op)
  friend type operator op(                                                     \
      const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &lhs,        \
      const uint16_t &rhs) {                                                   \
    (void)lhs;                                                                 \
    (void)rhs;                                                                 \
    throw runtime_error("joint matrix is not supported on host device.",       \
                        PI_INVALID_DEVICE);                                    \
  }                                                                            \
  friend type operator op(                                                     \
      const uint16_t &lhs,                                                     \
      const wi_element<uint16_t, NumRows, NumCols, Layout, Group> &rhs) {      \
    (void)lhs;                                                                 \
    (void)rhs;                                                                 \
    throw runtime_error("joint matrix is not supported on host device.",       \
                        PI_INVALID_DEVICE);                                    \
  }
  #endif // __SYCL_DEVICE_ONLY__
  OP(float, uint16_t, +)
  OP(float, uint16_t, -)
  OP(float, uint16_t, *)
  OP(float, uint16_t, /)
  OP(bool, bool, ==)
  OP(bool, bool, !=)
  OP(bool, bool, <)
  OP(bool, bool, >)
  OP(bool, bool, <=)
  OP(bool, bool, >=)
#undef OP
'''

Reply:
Yes, that's exactly what I was thinking.